pkg/runner/executable_path_resolving_runner.go

```diff
@@ -0,0 +1,121 @@
+package runner
+
+import (
+	"context"
+	"path/filepath"
+	"strings"
+
+	"github.com/buildbarn/bb-remote-execution/pkg/proto/runner"
+	"github.com/buildbarn/bb-storage/pkg/filesystem"
+	"github.com/buildbarn/bb-storage/pkg/util"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type rootDirectoryOpener func(inputRootDirectory string) (filesystem.DirectoryCloser, error)
+
+type executablePathResolvingRunner struct {
+	base                Runner
+	rootDirectoryOpener rootDirectoryOpener
+}
+
+// NewExecutablePathResolvingRunner modifies the arguments of the
+// incoming RunRequest so that the first argument is an absolute path
+// resolved using the path in the request, using the input root as the
+// chroot if necessary
+func NewExecutablePathResolvingRunner(base Runner, rootDirectoryOpener rootDirectoryOpener) Runner {
+	return &executablePathResolvingRunner{
+		base:                base,
+		rootDirectoryOpener: rootDirectoryOpener,
+	}
+}
+
+func (r *executablePathResolvingRunner) enterDir(file, chroot string) (filesystem.DirectoryCloser, error) {
+	rootDir, err := r.rootDirectoryOpener(chroot)
+	if err != nil {
+		return nil, util.StatusWrap(err, "Failed to open root directory")
+	}
+	components := strings.FieldsFunc(file, func(r rune) bool { return r == '/' })
+	dir := rootDir
+	for n, component := range components[:len(components)-1] {
+		dir2, err := dir.EnterDirectory(component)
+		if err != nil {
+			return nil, util.StatusWrapf(err, "Failed to enter directory %#v", filepath.Join(components[:n+1]...))
+		}
+		dir.Close()
+		dir = dir2
+	}
+	return dir, nil
+}
+
+func (r *executablePathResolvingRunner) findExecutable(file, chroot string) error {
+	dir, err := r.enterDir(file, chroot)
+	if err != nil {
+		return err
+	}
+	d, err := dir.Lstat(filepath.Base(file))
+	if err != nil {
+		return err
+	}
+	if m := d.Type(); m == filesystem.FileTypeExecutableFile {
+		return nil
+	} else if m == filesystem.FileTypeSymlink {
+		for i := 0; i < 10; i++ {
+			link, err := dir.Readlink(filepath.Base(file))
```

Comment:
What if you moved the top part of this loop until the Lstat() call to the bottom of the loop? I suspect that would allow you to remove some of the code above the loop?

A bit more explanation what I mean: Right now this entire function is somewhat like this:
```go
a
for {
  b
  a
}
```
But maybe it could be translated to:
```go
for {
  a
  b
}
```